[adjacent-overload-signatures] False positive on call signatures and a method named `call`

<!--
If you have a problem with a specific rule, please begin your issue title with [rulename] to make it easier to search for.
I.e. "[no-unused-vars] False positive when fooing the bar"

Please don't ignore this template.

If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
Are you opening an issue because the rule you're trying to use is not found?
🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨
1) Check the releases log: https://github.com/typescript-eslint/typescript-eslint/releases
    -  If the rule isn't listed there, then chances are it hasn't been released to the main npm tag yet.
2) Try installing the `canary` tag: `npm i @typescript-eslint/eslint-plugin@canary`.
    - The canary tag is built for every commit to master, so it contains the bleeding edge build.
3) If ESLint still can't find the rule, then consider reporting an issue.
-->

<!--
Make sure you read through our FAQ before posting.
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md
-->

**Repro**

<!--
Include a ***minimal*** reproduction case.
The more irrelevant code/config you give, the harder it is for us to investigate.
-->

```js
module.exports = {
  parser: "@typescript-eslint/parser",
  plugins: ["@typescript-eslint"],
  rules: {
    "@typescript-eslint/adjacent-overload-signatures": "error",
  },
};

```

```TS
interface Foo {
  (): void;
  n: number;
  call(): void;
}
```

**Expected Result**
No lint error is detected

**Actual Result**
```
  4:3  error  All 'call' signatures should be adjacent  @typescript-eslint/adjacent-overload-signatures

✖ 1 problem (1 error, 0 warnings)

```

**Additional Info**

<!--
Did eslint throw an exception?

Please run your lint again with the --debug flag, and dump the output below.
i.e. eslint --ext ".ts,.js" src --debug
-->

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.6.1` |
| `@typescript-eslint/parser`        | `3.6.1` |
| `TypeScript`                       | `3.9.5` |
| `ESLint`                           | `7.5.0` |
| `node`                             | `12.13.1` |
| `npm`                              | `6.12.1` |

--- a/packages/eslint-plugin/src/rules/adjacent-overload-signatures.ts
+++ b/packages/eslint-plugin/src/rules/adjacent-overload-signatures.ts
@@ -30,15 +30,23 @@ export default util.createRule({
   create(context) {
     const sourceCode = context.getSourceCode();
 
+    interface Method {
+      name: string;
+      static: boolean;
+      callSignature: boolean;
+    }
+
     /**
-     * Gets the name of the member being processed.
+     * Gets the name and attribute of the member being processed.
      * @param member the member being processed.
-     * @returns the name of the member or null if it's a member not relevant to the rule.
+     * @returns the name and attribute of the member or null if it's a member not relevant to the rule.
      */
-    function getMemberName(member: TSESTree.Node): string | null {
+    function getMemberMethod(member: TSESTree.Node): Method | null {
       if (!member) {
         return null;
       }
+
+      const isStatic = 'static' in member && !!member.static;
 
       switch (member.type) {
         case AST_NODE_TYPES.ExportDefaultDeclaration:
@@ -49,33 +57,55 @@ export default util.createRule({
             return null;
           }
 
-          return getMemberName(member.declaration);
+          return getMemberMethod(member.declaration);
         }
         case AST_NODE_TYPES.TSDeclareFunction:
-        case AST_NODE_TYPES.FunctionDeclaration:
-          return member.id?.name ?? null;
+        case AST_NODE_TYPES.FunctionDeclaration: {
+          const name = member.id?.name ?? null;
+          if (name === null) {
+            return null;
+          }
+          return {
+            name,
+            static: isStatic,
+            callSignature: false,
+          };
+        }
         case AST_NODE_TYPES.TSMethodSignature:
-          return util.getNameFromMember(member, sourceCode);
+          return {
+            name: util.getNameFromMember(member, sourceCode),
+            static: isStatic,
+            callSignature: false,
+          };
         case AST_NODE_TYPES.TSCallSignatureDeclaration:
-          return 'call';
+          return {
+            name: 'call',
+            static: isStatic,
+            callSignature: true,
+          };
         case AST_NODE_TYPES.TSConstructSignatureDeclaration:
-          return 'new';
+          return {
+            name: 'new',
+            static: isStatic,
+            callSignature: false,
+          };
         case AST_NODE_TYPES.MethodDefinition:
-          return util.getNameFromMember(member, sourceCode);
+          return {
+            name: util.getNameFromMember(member, sourceCode),
+            static: isStatic,
+            callSignature: false,
+          };
       }
 
       return null;
     }
 
-    interface Method {
-      name: string;
-      static: boolean;
-    }
     function isSameMethod(method1: Method, method2: Method | null): boolean {
       return (
         !!method2 &&
         method1.name === method2.name &&
-        method1.static === method2.static
+        method1.static === method2.static &&
+        method1.callSignature === method2.callSignature
       );
     }
 
@@ -104,15 +134,11 @@ export default util.createRule({
         const seenMethods: Method[] = [];
 
         members.forEach(member => {
-          const name = getMemberName(member);
-          if (name === null) {
+          const method = getMemberMethod(member);
+          if (method === null) {
             lastMethod = null;
             return;
           }
-          const method = {
-            name,
-            static: 'static' in member && !!member.static,
-          };
 
           const index = seenMethods.findIndex(seenMethod =>
             isSameMethod(method, seenMethod),

--- a/packages/eslint-plugin/tests/rules/adjacent-overload-signatures.test.ts
+++ b/packages/eslint-plugin/tests/rules/adjacent-overload-signatures.test.ts
@@ -132,6 +132,17 @@ interface Foo {
     `,
     `
 interface Foo {
+  (s: string): void;
+  (n: number): void;
+  (sn: string | number): void;
+  foo(n: number): void;
+  bar(): void;
+  baz(): void;
+  call(): void;
+}
+    `,
+    `
+interface Foo {
   foo(s: string): void;
   foo(n: number): void;
   foo(sn: string | number): void;
@@ -534,6 +545,7 @@ interface Foo {
   (sn: string | number): void;
   bar(): void;
   baz(): void;
+  call(): void;
 }
       `,
       errors: [

